src/verifast1.ml

```diff
@@ -3189,18 +3255,72 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
           static_error l (match language with CLang -> "No such function: " ^ g | Java -> "No such method or function: " ^ g) None
       in
       if language = CLang || classmap = [] then func_call () else
-      let try_qualified_call tn es args fb on_fail =
+      let try_qualified_call tn es args fb real_type_arguments on_fail =
         let ms = get_methods tn g in
         if ms = [] then on_fail () else
         let argtps = List.map (fun e -> let (_, tp, _) = (check e) in tp) args in
+        let ms = List.map (fun (sign, (tn', lm, gh, rt, xmap, pre, post, epost, terminates, fb', v, abstract, tparams)) ->
+            (* Build a list of type arguments, either from the ones that got passed, or the ones you have to assume *)
+            let targenv = 
+              (if gh == Real && (List.length real_type_arguments) > 0 
+                then begin
+                let type_arguments_count = List.length real_type_arguments in
+                let tparams_count = List.length tparams in
+                if type_arguments_count <> tparams_count
+                  then static_error l 
+                    (Printf.sprintf "Wrong amount of type arguments provided for method: %s. Expected: %s Actual: %s" 
+                      g
+                      (string_of_int tparams_count) 
+                      (string_of_int type_arguments_count)
+                     ) None
+                  else List.map2 (fun a b -> (a,b)) tparams real_type_arguments 
+                end
+                else (*Build a list of type arguments, inferred from the signature *)

```

Comment:
Perhaps it's better to not try to do any type argument inference in this PR, and leave it for another PR?